src/poly/commitment.rs

```diff
@@ -154,6 +245,52 @@ impl<C: CurveAffine> Params<C> {
     }
 }
 
+/// A guard returned by the verifier
+#[derive(Debug)]
+pub struct Guard<C: CurveAffine> {
+    msm: MSM<C>,
+    neg_z1: C::Scalar,
+    allinv: C::Scalar,
+    challenges_sq: Vec<C::Scalar>,
+    challenges_sq_packed: Vec<Challenge>,
+}
+
+impl<C: CurveAffine> Guard<C> {
+    /// Lets caller supply the challenges and obtain an MSM with updated
+    /// scalars and points.
+    pub fn use_challenges(mut self, params: &Params<C>) -> Result<MSM<C>, Error> {
+        let mut scalars: Vec<C::Scalar> = vec![];
```

Comment:
We shouldn't need to keep a `scalars`/`bases` temporaries around here. The MSM will do everything for us.

We just need to do

```rust
        // - [z1] G
        let s = compute_s(&self.challenges_sq, self.allinv * &self.neg_z1);
        self.msm.add_to_g(&s);

        self.msm
```